File: src/java/src/main/java/com/maimemo/PhraseSearchTree.java
package com.maimemo;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by TJT on 3/27/16.
 */
class PhraseSearchTree {

    private final Pattern REPLACEMENT_REGEX = Pattern.compile(" *(\\boneself\\b|\\bsomebody's\\b|\\bsomebody\\b|\\bsomething\\b|" +
            "\\bone's\\b|\\bdo sth\\.|\\bsb's\\b|\\bdo sth\\b|" +
            "\\bsb\\.'s\\b|\\bsb\\.|\\bsb\\b|\\bsth\\.|\\bsth\\b|\\.\\.\\.) *", Pattern.CASE_INSENSITIVE);

    private final Pattern SPLIT_SPACE_REGEX = Pattern.compile(" ");

    private final Matcher ANY_MATCHER = REPLACEMENT_REGEX.matcher("");

    private static final int BASE_SIZE = 800;

    private final CharSequenceWrapper ANY = new CharSequenceWrapper("");

    private CharSequence[] library;

    private Node root;

    private volatile boolean builded = false;

    PhraseSearchTree(CharSequence[] library) {
        this.library = library;
    }

    private boolean isAny(CharSequence word) {
        ANY_MATCHER.reset(word);
        return ANY_MATCHER.find();
    }

    private class BuilderNode {
        private int hash = 0;
        public List<CharSequence> words = new LinkedList<>();
        public List<CharSequence> array;

        @Override
        public int hashCode() {
            int h = hash;
            if (h != 0 && array.size() > 0) {
                for (CharSequence s : array) {
                    for (int i = 0; i < s.length(); i++) {
                        h = 31 * h + s.charAt(i);
                    }
                }
            }
            return hash;
        }

        @Override
        public boolean equals(Object obj) {
            if (obj instanceof BuilderNode) {
                BuilderNode builderNode = (BuilderNode) obj;
                if (builderNode.array.size() != array.size()) {
                    return false;
                }
                for (int i = 0; i < array.size(); i++) {
                    if (!builderNode.array.get(i).equals(array.get(i))) {
                        return false;
                    }
                }
                return true;
            }
            return false;
        }
    }

    public void build() {
        Map<BuilderNode, BuilderNode> builderNodeSet = new HashMap<>(BASE_SIZE);

        root = new Node();
        root.children = new HashMap<>(BASE_SIZE / 4);

        for (CharSequence s : library) {
            if (TextUtils.indexOf(s, ' ', 0) == -1) {
                continue;
            }
            if (isAny(s)) {
                CharSequence[] _split = SPLIT_SPACE_REGEX.split(s);
                List<CharSequence> _split1 = new LinkedList<>();
                for (int i = 0; i < _split.length; i++) {
                    if (isAny(_split[i])) {
                        if (i != _split.length - 1) {
                            _split1.add(ANY);
                        }
                    } else {
                        _split1.add(new CharSequenceWrapper(_split[i]));
                    }
                }

                BuilderNode node = new BuilderNode();
                node.words.add(s);
                node.array = _split1;

                BuilderNode n = builderNodeSet.get(node);
                if (n != null) {
                    n.words.add(s);
                } else {
                    builderNodeSet.put(node, node);
                }
            }
        }

        for (BuilderNode n : builderNodeSet.values()) {
            addWordToTree(n);
        }

        // move any nodes of root node to children of root
        if (root.any != null) {
            for (Node node : root.any) {
                if (node.children != null) {
                    root.children.putAll(node.children);
                }
            }
        }
        root.any = null;

        builded = true;
    }

    private void addWordToTree(BuilderNode node) {
        List<CharSequence> splittedWords = node.array;
        int size = splittedWords.size();
        Node parent = root;
        Node child;
        for (int i = 0; i < size; i++) {
            child = null;
            if (parent.children != null) {
                child = parent.children.get(splittedWords.get(i));
            }
            if (child == null) {
                child = new Node();
                if (i == size - 1) {
                    child.word = node.words;
                }
                // is any
                if (splittedWords.get(i) == ANY) {
                    if (parent.any == null) {
                        parent.any = new LinkedList<>();
                    }
                    parent.any.add(child);
                } else {
                    if (parent.children == null) {
                        parent.children = new TreeMap<>();
                    }
                    parent.children.put(splittedWords.get(i), child);
                }
            }
            parent = child;
        }
    }

    public boolean isBuilded() {
        return builded;
    }

    private final SubCharSequence subCharSequence = new SubCharSequence();
    private final FastIntPairArray intPairArray = new FastIntPairArray(20);
    private final WordIterator wordIterator = new WordIterator();

    public List<CharSequence> search(CharSequence sentence) {
        wordIterator.update(sentence);
        wordIterator.setEnableSplitHyphen(false);
        wordIterator.justSplit(intPairArray);
        int size = intPairArray.size();
        if (size < 2) {
            return null;
        }
        List<CharSequence> word = null;

        for (int i = 0; i < size - 2; i++) {
            List<CharSequence> _word = _search(sentence, intPairArray, i);
            if (_word != null) {
                if (word == null) {
                    word = new LinkedList<>();
                }
                word.addAll(_word);
            }
        }

        return word;
    }

    private List<CharSequence> _search(CharSequence sentence, FastIntPairArray intPairArray, int start) {
        int size = intPairArray.size();
        List<CharSequence> word = null;
        Node node = root;
        for (int i = start; i < size; i++) {
            subCharSequence.update(sentence, intPairArray.getStart(i), intPairArray.getEnd(i));

            if (node.any != null) {
                Node nn;
                for (Node n : node.any) {
                    if (n.children != null) {
                        nn = n.children.get(subCharSequence);
                        if (nn != null) {
                            node = nn;
                            break;
                        }
                    }
                }
            }

            if (node.word != null) {
                word = node.word;
                break;
            }
            Node n = null;
            if (node.children != null) {
                n = node.children.get(subCharSequence);
            }
            if (n != null) {
                node = n;
            }
        }
        return word;
    }

    private class Node {
        public List<CharSequence> word;
        public CharSequence[] splittedWords;
        public List<Node> any;
        public Map<CharSequence, Node> children;
    }
}

File: src/WordExtractor/WordExtractor/PhraseSearchTree.m
                                                                     //
//  PhraseSearchTree.m
//  WordExtractor
//
//  Created by TJT on 8/19/16.
//  Copyright © 2016 TJT. All rights reserved.
//

#import "PhraseSearchTree.h"
#import "TextUtil.h"
#import "StringWrapper.h"
#import "SubString.h"
#import "TextIterators.h"

@interface BuilderNode : NSObject<NSCopying>

@property (nonatomic, assign) NSUInteger                    hashCache;
@property (nonatomic, strong) NSMutableArray<NSString *>    *words;
@property (nonatomic, strong) NSArray<NSString *>           *array;

@end

@implementation BuilderNode

- (id)copyWithZone:(NSZone *)zone {
    return self;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _words = [NSMutableArray new];
        _hashCache = 0;
    }
    return self;
}

- (NSUInteger)hash {
    if (_hashCache != 0 && nil != _array && _array.count > 0) {
        for (NSString *s in _array) {
            for (NSInteger i = 0; i < s.length; i++) {
                _hashCache = _hashCache * 31 + [s characterAtIndex:i];
            }
        }
    }
    return _hashCache;
}

- (BOOL)isEqual:(id)other {
    if (other == self) {
        return YES;
    } else {
        if ([other isKindOfClass:[self class]]) {
            BuilderNode *n = other;
            if (n.array.count != _array.count) {
                return NO;
            }
            for (NSInteger i = 0; i < _array.count; i++) {
                if (![n.array[i] isEqual:_array[i]]) {
                    return NO;
                }
            }
            return YES;
        }
        return NO;
    }
}

@end

@interface Node : NSObject

@property (nonatomic, strong) NSArray<NSString *>              *word;
@property (nonatomic, strong) NSMutableArray<Node *>           *any;
@property (nonatomic, strong) NSMutableDictionary<NSString *, Node *>   *children;

@end

@implementation Node

- (instancetype)init {
    self = [super init];
    return self;
}

@end

@interface PhraseSearchTree ()

@property (nonatomic, strong) NSRegularExpression       *ANY_PATTERN;
@property (nonatomic, strong) Node                      *root;
@property (nonatomic, strong) NSArray<NSString *>       *library;
@property (nonatomic, strong) RangeArray                *splitResult;
@property (nonatomic, strong) SubString                 *subString;
@property (nonatomic, strong) WordIterator              *wordIterator;

@end

@implementation PhraseSearchTree

StringWrapper *ANY;

- (instancetype)initWithLibrary:(NSArray<NSString *> *)library {
    ANY = [StringWrapper stringWithString:@""];
    _ANY_PATTERN = [NSRegularExpression regularExpressionWithPattern:@" *(\\boneself\\b|\\bsomebody's\\b|\\bsomebody\\b|\\bsomething\\b|\\bone's\\b|\\bdo sth\\.|\\bsb's\\b|\\bdo sth\\b|\\bsb\\.'s\\b|\\bsb\\.|\\bsb\\b|\\bsth\\.|\\bsth\\b|\\.\\.\\.) *" options:NSRegularExpressionCaseInsensitive error:nil];
    _splitResult = [RangeArray new];
    _library = library;
    _isBuilded = NO;
    _wordIterator = [WordIterator new];
    _subString = [SubString new];
    return self;
}

- (BOOL)isAny:(NSString *)word {
    return [_ANY_PATTERN numberOfMatchesInString:word options:0 range:NSMakeRange(0, word.length)] > 0;
}

- (void)addNodeToTree:(BuilderNode *)node {
    NSArray *splittedWords = node.array;
    NSInteger size = splittedWords.count;
    Node *parent = _root;
    Node *child;
    for (NSInteger i = 0; i < size; i++) {
        child = nil;
        if (nil != parent.children) {
            child = [parent.children objectForKey:splittedWords[i]];
        }
        if (nil == child) {
            child = [Node new];
            if (i == size - 1) {
                child.word = node.words;
            }
            // is any
            if (splittedWords[i] == ANY) {
                if (nil == parent.any) {
                    parent.any = [NSMutableArray new];
                }
                [parent.any addObject:child];
            } else {
                if (nil == parent.children) {
                    parent.children = [NSMutableDictionary new];
                }
                [parent.children setObject:child forKey:splittedWords[i]];
            }
        }
        parent = child;
    }
}

- (void)build {
    NSMutableDictionary *builderNodeSet = [NSMutableDictionary new];
    _root = [Node new];
    _root.children = [NSMutableDictionary new];
    
    for (NSString *voc in _library) {
        if (NSNotFound == [voc rangeOfString:@" "].location) {
            continue;
        }
        
        if ([self isAny:voc]) {
            [_splitResult clear];
            [TextUtil split:voc separator:@" " outValue:_splitResult];
            NSMutableArray *split = [NSMutableArray new];
            for (NSInteger i = 0; i < _splitResult.count; i++) {
                NSString *s = [voc substringWithRange:NSMakeRange([_splitResult startAtIndex:i], [_splitResult endAtIndex:i] - [_splitResult startAtIndex:i])];
                if ([self isAny:s]) {
                    if (i != _splitResult.count - 1) {
                        [split addObject:ANY];
                    }
                } else {
                    [split addObject:[StringWrapper stringWithString:s]];
                }
            }
            
            BuilderNode *node = [BuilderNode new];
            [node.words addObject:voc];
            node.array = split;
            
            BuilderNode *n = [builderNodeSet objectForKey:node];
            if (nil != n) {
                [n.words addObject:voc];
            } else {
                [builderNodeSet setObject:node forKey:node];
            }
        }
    }
    for (BuilderNode *n in builderNodeSet) {
        [self addNodeToTree:n];
    }
    
    if (nil != _root.any) {
        for (Node *n in _root.any) {
            if (nil != n.children) {
                [_root.children setValuesForKeysWithDictionary:n.children];
            }
        }
    }
    _root.any = nil;
    self.isBuilded = YES;
}

- (NSArray *)search:(NSString *)sentence split:(RangeArray *)split start:(NSInteger)start {
    NSInteger size = split.count;
    NSArray<NSString *> *word = nil;
    Node *node = _root;
    for (NSInteger i = start; i < size; i++) {
        [_subString updateWithString:sentence start:[split startAtIndex:i] end:[split endAtIndex:i]];
        
        if (nil != node.any) {
            Node *nn;
            for (Node *n in node.any) {
                if (nil != n.children) {
                    nn = [n.children objectForKey:_subString];
                }
                if (nil != nn) {
                    node = nn;
                    break;
                }
            }
        }
        
        if (nil != node.word) {
            word = node.word;
            break;
        }
        
        Node *n = nil;
        if (nil != node.children) {
            n = [node.children objectForKey:_subString];
        }
        if (nil != n) {
            node = n;
        }
    }
    return word;
}

- (NSArray *)search:(NSString *)sentence {
    [_wordIterator update:sentence];
    _wordIterator.isSplitHyphen = NO;
    [_wordIterator justSplit:_splitResult];
    NSInteger size = _splitResult.count;
    if (size < 2) {
        return nil;
    }
    
    NSMutableArray<NSString *> *word = nil;
    
    for (NSInteger i = 0; i < size - 2; i++) {
        NSArray<NSString *> *words = [self search:sentence split:_splitResult start:i];
        if (nil != words) {
            if (nil == word) {
                word = [NSMutableArray new];
            }
            [word addObjectsFromArray:words];
        }
    }
    return word;
}

@end
